pkg/streamingpromql/operators/aggregations/count_values.go

```diff
@@ -0,0 +1,247 @@
+			}
+		}
+
+		for _, p := range data.Histograms {
+			if err := c.incrementCount(s.Labels, p.T, p.H.String(), accumulator); err != nil {
+				return nil, err
+			}
+		}
+
+		types.PutInstantVectorSeriesData(data, c.MemoryConsumptionTracker)
+	}
+
+	outputMetadata := types.GetSeriesMetadataSlice(len(accumulator))
+	c.series = make([][]promql.FPoint, 0, len(accumulator))
+
+	for _, s := range accumulator {
+		outputMetadata = append(outputMetadata, types.SeriesMetadata{Labels: s.labels})
+
+		points, err := s.ToPoints(c.MemoryConsumptionTracker, c.TimeRange)
+		if err != nil {
+			return nil, err
+		}
+
+		c.series = append(c.series, points)
+
+		types.IntSlicePool.Put(s.count, c.MemoryConsumptionTracker)
+		s.count = nil
+		countValuesSeriesPool.Put(s)
+	}
+
+	return outputMetadata, nil
+}
+
+func (c *CountValues) loadLabelName() error {
+	c.resolvedLabelName = c.LabelName.GetValue()
+	if !model.LabelName(c.resolvedLabelName).IsValid() {
+		return fmt.Errorf("invalid label name %q for count_values", c.resolvedLabelName)
+	}
+
+	if c.Without {
+		c.Grouping = append(c.Grouping, labels.MetricName)
+	}
+
+	slices.Sort(c.Grouping)
+	return nil
+}
+
+func (c *CountValues) formatFloatValue(f float64) string {
+	// Using AppendFloat like this (rather than FormatFloat) allows us to reuse the buffer used for formatting the string,
+	// rather than creating a new one for every value.
+	c.valueBuffer = c.valueBuffer[:0]
+	c.valueBuffer = strconv.AppendFloat(c.valueBuffer, f, 'f', -1, 64)
+
+	return string(c.valueBuffer)
+}
+
+func (c *CountValues) incrementCount(seriesLabels labels.Labels, t int64, value string, accumulator map[string]*countValuesSeries) error {
+	l := c.computeOutputLabels(seriesLabels, value)
+	c.labelsBytesBuffer = l.Bytes(c.labelsBytesBuffer)
+	series, exists := accumulator[string(c.labelsBytesBuffer)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+	if !exists {
+		series = countValuesSeriesPool.Get().(*countValuesSeries)
+		series.labels = l
+		series.outputPointCount = 0
+
+		var err error
+		series.count, err = types.IntSlicePool.Get(c.TimeRange.StepCount, c.MemoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+
+		series.count = series.count[:c.TimeRange.StepCount]
+		accumulator[string(c.labelsBytesBuffer)] = series
+	}
+
+	idx := c.TimeRange.PointIndex(t)
+
+	if series.count[idx] == 0 {
+		series.outputPointCount++
+	}
+
+	series.count[idx]++
+
+	return nil
+}
+
+func (c *CountValues) computeOutputLabels(seriesLabels labels.Labels, value string) labels.Labels {
+	c.labelsBuilder.Reset(seriesLabels)
+
+	if c.Without {
+		c.labelsBuilder.Del(c.Grouping...)
+	} else {
+		c.labelsBuilder.Keep(c.Grouping...)
+	}
+
+	c.labelsBuilder.Set(c.resolvedLabelName, value)
+
+	return c.labelsBuilder.Labels()
+}
+
+func (s *countValuesSeries) ToPoints(memoryConsumptionTracker *limiting.MemoryConsumptionTracker, timeRange types.QueryTimeRange) ([]promql.FPoint, error) {
```

Comment:
Is there any reason to make this method public?

Reply:
Good catch, fixed in 13fbbe8